In [60]:
#---Step 0: Import necessary libraries and render plots inline---

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

plt.style.use('fivethirtyeight')

In [61]:
user_usage = pd.read_csv("/Users/JCheong/Merge_Join_Data/user_usage.csv")
user_device = pd.read_csv("/Users/JCheong/Merge_Join_Data/user_device.csv")
devices = pd.read_csv("/Users/JCheong/Merge_Join_Data/android_devices.csv")

In [62]:
user_usage.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
0                    21.97                    4.82     1557.33   22787
1                  1710.08                  136.88     7267.55   22788
2                  1710.08                  136.88     7267.55   22789
3                    94.46                   35.17      519.12   22790
4                    71.59                   79.26     1557.33   22792

In [63]:
user_device.head()

use_id  user_id platform  platform_version     device  use_type_id
0   22782    26980      ios              10.2  iPhone7,2            2
1   22783    29628  android               6.0    Nexus 5            3
2   22784    28473  android               5.1   SM-G903F            1
3   22785    15200      ios              10.2  iPhone7,2            3
4   22786    28239  android               6.0  ONE E1003            1

In [64]:
android_devices.head()

Retail Branding Marketing Name    Device                      Model
0             NaN            NaN    AD681H  Smartfren Andromax AD681H
1             NaN            NaN     FJL21                      FJL21
2             NaN            NaN       T31              Panasonic T31
3             NaN            NaN  hws7721g         MediaPad 7 Youth 2
4              3Q        OC1020A   OC1020A                    OC1020A

There are linking attributes between the sample datasets that are important to note – “use_id” is shared between the user_usage and user_device, and the “device” column of user_device and “Model” column of the devices dataset contain common codes.

# Sample Problem

We would like to determine if the usage patterns for users differ between different devices. For example, do users using Samsung devices use more call minutes than those using  LG devices?

We want to form a single dataframe with columns for user usage figures (calls per month, sms per month etc) and also columns with device information (model, manufacturer, etc).

We will need to “merge” (or “join”) our sample datasets together into one single dataset for analysis.

# Merging Dataframes

“Merging” two datasets is the process of bringing two datasets together into one, and aligning the rows from each based on common attributes or columns.

1. For each row in the user_usage dataset – make a new column that contains the “device” code from the user_devices dataframe. i.e. for the first row, the use_id is 22787, so we go to the user_devices dataset, find the use_id 22787, and copy the value from the “device” column across.

2. After this is complete, we take the new device columns, and we find the corresponding “Retail Branding” and “Model” from the devices dataset.

3. Finally, we can look at different statistics for usage splitting and grouping data by the device manufacturers used.

In [65]:
# Use the default inner merge to combine user_usage and user_device based on use_id
first_merge = pd.merge(user_usage, user_device[["use_id", "platform", "device"]], on="use_id")

first_merge.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device  
0  android  GT-I9505  
1  android  SM-G930F  
2  android  SM-G930F  
3  android     D2303  
4  android  SM-G361F

There are three different types of merges available in Pandas. These merge types are common across most database and data-orientated languages (SQL, R, SAS) and are typically referred to as “joins”.

1. Inner Merge / Inner join – The default Pandas behaviour, only keep rows where the merge “on” value exists in both the left and right dataframes.

2. Left Merge / Left outer join – (aka left merge or left join) Keep every row in the left dataframe. Where there are missing values of the “on” variable in the right dataframe, add empty / NaN values in the result.

3. Right Merge / Right outer join – (aka right merge or right join) Keep every row in the right dataframe. Where there are missing values of the “on” variable in the left column, add empty / NaN values in the result.

4. Outer Merge / Full outer join – A full outer join returns all the rows from the left dataframe, all the rows from the right dataframe, and matches up rows where possible, with NaNs elsewhere.

The merge type to use is specified using the “how” parameter in the merge command, taking values “left”, “right”, “inner” (default), or “outer”.

# Example of Left Merge / Left Join

In [66]:
left_merge_test = pd.merge(user_usage,
                           user_device[["use_id", "platform", "device"]],
                           on="use_id",
                           how="left"
                          )

print("There are {} missing values in the result.".format(left_merge_test["device"].isnull().sum()))

There are 81 missing values in the result.


In [67]:
left_merge_test

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                      21.97                    4.82     1557.33   22787   
1                    1710.08                  136.88     7267.55   22788   
2                    1710.08                  136.88     7267.55   22789   
3                      94.46                   35.17      519.12   22790   
4                      71.59                   79.26     1557.33   22792   
5                      71.59                   79.26     1557.33   22793   
6                      71.59                   79.26      519.12   22794   
7                      71.59                   79.26      519.12   22795   
8                      30.92                   22.77     3114.67   22799   
9                      69.80                   14.70    25955.55   22801   
10                    554.41                  150.06     3114.67   22804   
11                    189.10                   24.08      519.12   22805   
12                    283.30                  107.47    15573.33   22806   
13                    324.34                   92.52      519.12   22808   
14                    797.06                    7.67      519.12   22813   
15                    797.06                    7.67    15573.33   22814   
16                    797.06                    7.67    15573.33   22815   
17                    797.06                    7.67    15573.33   22816   
18                    797.06                    7.67    15573.33   22817   
19                     78.80                  327.33    10382.21   22819   
20                     78.80                  327.33    15573.33   22820   
21                     78.80                  327.33    15573.33   22822   
22                    164.10                  192.64     3114.67   22823   
23                    208.26                   91.76     5191.12   22824   
24                    681.44                   47.35     1271.39   22829   
25                    324.27                   91.50      519.12   22830   
26                     85.97                   26.94      407.01   22831   
27                    244.88                  105.95     1557.33   22832   
28                    135.09                   42.02     5191.12   22833   
29                     57.49                   16.73    15573.33   22839   
..                       ...                     ...         ...     ...   
210                   677.02                  101.86     1132.23   24318   
211                   677.02                  101.86     1132.23   24319   
212                   583.73                  216.52     1441.29   24332   
213                   583.73                  216.52     1441.29   24333   
214                   333.11                   70.91     2814.64   24399   
215                   677.02                  101.86     1132.23   24436   
216                   677.02                  101.86     1132.23   24437   
217                   677.02                  101.86     1132.23   24438   
218                   677.02                  101.86     1132.23   24439   
219                   463.05                  165.89      362.02   24445   
220                   205.40                   42.08     1045.39   24451   
221                   205.40                   42.08     1045.39   24455   
222                   436.37                  180.33      415.10   24466   
223                   583.73                  216.52     1441.29   24545   
224                   512.48                   61.63     4565.33   24601   
225                   291.82                  109.26     4527.83   24602   
226                   234.70                  167.39     2609.20   24647   
227                   291.82                  109.26     4527.83   24653   
228                  1083.14                   52.03     6215.01   24728   
229                   248.56                  133.07     2004.88   24761   
230                   583.73                  216.52 

# Outer Merge / Full Outer Join

Finally, we will perform an outer merge using Pandas, also referred to as a “full outer join” or just “outer join”. An outer join can be seen as a combination of left and right joins, or the opposite of an inner join. In outer joins, every row from the left and right dataframes is retained in the result, with NaNs where there are no matched join variables.

As such, we would expect the results to have the same number of rows as there are distinct values of “use_id” between user_device and user_usage, i.e. every join value from the left dataframe will be in the result along with every value from the right dataframe, and they’ll be linked where possible.

In [68]:
unique_use_ids = pd.concat([user_usage["use_id"], user_device["use_id"]], axis=0).unique()

print("There are {} unique values of use_id in our dataframe.".format(unique_use_ids.shape[0]))

outer_merge_test = pd.merge(user_usage,
                            user_device[["use_id", "platform", "device"]],
                            on="use_id",
                            how="outer"
                           )

outer_merge_test

There are 353 unique values of use_id in our dataframe.


outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                      21.97                    4.82     1557.33   22787   
1                    1710.08                  136.88     7267.55   22788   
2                    1710.08                  136.88     7267.55   22789   
3                      94.46                   35.17      519.12   22790   
4                      71.59                   79.26     1557.33   22792   
5                      71.59                   79.26     1557.33   22793   
6                      71.59                   79.26      519.12   22794   
7                      71.59                   79.26      519.12   22795   
8                      30.92                   22.77     3114.67   22799   
9                      69.80                   14.70    25955.55   22801   
10                    554.41                  150.06     3114.67   22804   
11                    189.10                   24.08      519.12   22805   
12                    283.30                  107.47    15573.33   22806   
13                    324.34                   92.52      519.12   22808   
14                    797.06                    7.67      519.12   22813   
15                    797.06                    7.67    15573.33   22814   
16                    797.06                    7.67    15573.33   22815   
17                    797.06                    7.67    15573.33   22816   
18                    797.06                    7.67    15573.33   22817   
19                     78.80                  327.33    10382.21   22819   
20                     78.80                  327.33    15573.33   22820   
21                     78.80                  327.33    15573.33   22822   
22                    164.10                  192.64     3114.67   22823   
23                    208.26                   91.76     5191.12   22824   
24                    681.44                   47.35     1271.39   22829   
25                    324.27                   91.50      519.12   22830   
26                     85.97                   26.94      407.01   22831   
27                    244.88                  105.95     1557.33   22832   
28                    135.09                   42.02     5191.12   22833   
29                     57.49                   16.73    15573.33   22839   
..                       ...                     ...         ...     ...   
323                      NaN                     NaN         NaN   22976   
324                      NaN                     NaN         NaN   22983   
325                      NaN                     NaN         NaN   22984   
326                      NaN                     NaN         NaN   22990   
327                      NaN                     NaN         NaN   22993   
328                      NaN                     NaN         NaN   22996   
329                      NaN                     NaN         NaN   23000   
330                      NaN                     NaN         NaN   23001   
331                      NaN                     NaN         NaN   23004   
332                      NaN                     NaN         NaN   23006   
333                      NaN                     NaN         NaN   23007   
334                      NaN                     NaN         NaN   23008   
335                      NaN                     NaN         NaN   23009   
336                      NaN                     NaN         NaN   23010   
337                      NaN                     NaN         NaN   23011   
338                      NaN                     NaN         NaN   23014   
339                      NaN                     NaN         NaN   23022   
340                      NaN                     NaN         NaN   23025   
341                      NaN                     NaN         NaN   23033   
342                      NaN                     NaN         NaN   23034   
343                      NaN                     NaN 

In [69]:
NaN_counts = pd.DataFrame(outer_merge_test.apply(lambda x: x.isnull().sum(), axis=1))
counter = (NaN_counts[0] == 0).sum()

print("There are {} rows with no missing values.".format(counter))

There are 159 rows with no missing values.


# Using Indicator Parameter to Track Merges

To assist with the identification of where rows originate from, Pandas provides an “indicator” parameter that can be used with the merge function which creates an additional column called “_merge” in the output that labels the original source for each row.

In [70]:
indicator_test = pd.merge(user_usage,
                          user_device[["use_id", "platform", "device"]],
                          on="use_id",
                          how="outer",
                          indicator=True
                         )

indicator_test

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                      21.97                    4.82     1557.33   22787   
1                    1710.08                  136.88     7267.55   22788   
2                    1710.08                  136.88     7267.55   22789   
3                      94.46                   35.17      519.12   22790   
4                      71.59                   79.26     1557.33   22792   
5                      71.59                   79.26     1557.33   22793   
6                      71.59                   79.26      519.12   22794   
7                      71.59                   79.26      519.12   22795   
8                      30.92                   22.77     3114.67   22799   
9                      69.80                   14.70    25955.55   22801   
10                    554.41                  150.06     3114.67   22804   
11                    189.10                   24.08      519.12   22805   
12                    283.30                  107.47    15573.33   22806   
13                    324.34                   92.52      519.12   22808   
14                    797.06                    7.67      519.12   22813   
15                    797.06                    7.67    15573.33   22814   
16                    797.06                    7.67    15573.33   22815   
17                    797.06                    7.67    15573.33   22816   
18                    797.06                    7.67    15573.33   22817   
19                     78.80                  327.33    10382.21   22819   
20                     78.80                  327.33    15573.33   22820   
21                     78.80                  327.33    15573.33   22822   
22                    164.10                  192.64     3114.67   22823   
23                    208.26                   91.76     5191.12   22824   
24                    681.44                   47.35     1271.39   22829   
25                    324.27                   91.50      519.12   22830   
26                     85.97                   26.94      407.01   22831   
27                    244.88                  105.95     1557.33   22832   
28                    135.09                   42.02     5191.12   22833   
29                     57.49                   16.73    15573.33   22839   
..                       ...                     ...         ...     ...   
323                      NaN                     NaN         NaN   22976   
324                      NaN                     NaN         NaN   22983   
325                      NaN                     NaN         NaN   22984   
326                      NaN                     NaN         NaN   22990   
327                      NaN                     NaN         NaN   22993   
328                      NaN                     NaN         NaN   22996   
329                      NaN                     NaN         NaN   23000   
330                      NaN                     NaN         NaN   23001   
331                      NaN                     NaN         NaN   23004   
332                      NaN                     NaN         NaN   23006   
333                      NaN                     NaN         NaN   23007   
334                      NaN                     NaN         NaN   23008   
335                      NaN                     NaN         NaN   23009   
336                      NaN                     NaN         NaN   23010   
337                      NaN                     NaN         NaN   23011   
338                      NaN                     NaN         NaN   23014   
339                      NaN                     NaN         NaN   23022   
340                      NaN                     NaN         NaN   23025   
341                      NaN                     NaN         NaN   23033   
342                      NaN                     NaN         NaN   23034   
343                      NaN                     NaN 

# Final Merge - Joining Device Details to Result

Coming back to our original problem, we have already merged user_usage with user_device, so we have the platform and device for each user. Originally, we used an “inner merge” as the default in Pandas, and as such, we only have entries for users where there is also device information. We’ll redo this merge using a left join to keep all users, and then use a second left merge to finally to get the device manufacturers in the same dataframe.

In [71]:
# First, add the platform and device to the user usage using a left join
result = pd.merge(user_usage,
                  user_device[["use_id", "platform", "device"]],
                  on="use_id",
                  how="left"
                 )

# Second, add manufacturer and model columns based on "device column" from the previous dataframe and the "model" column in the device dataframe
devices.rename(columns={"Retail Branding" : "manufacturer"}, inplace=True)
result = pd.merge(result,
                  devices[["manufacturer","Model"]],
                  left_on="device",
                  right_on="Model",
                  how="left"
                 )

result.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device manufacturer     Model  
0  android  GT-I9505      Samsung  GT-I9505  
1  android  SM-G930F      Samsung  SM-G930F  
2  android  SM-G930F      Samsung  SM-G930F  
3  android     D2303         Sony     D2303  
4  android  SM-G361F      Samsung  SM-G361F

The columns used in a merge operator do not need to be named the same in both the left and right dataframe. In the second merge above, note that the device ID is called “device” in the left dataframe, and called “Model” in the right dataframe.

Different column names are specified for merges in Pandas using the “left_on” and “right_on” parameters, instead of using only the “on” parameter.

# Calculating Statistics Based on Device

With our merges complete, we can use the data aggregation functionality of Pandas to quickly work out the mean usage for users based on device manufacturer. Note that the small sample size creates even smaller groups, so I wouldn’t attribute any statistical significance to these particular results.

In [79]:
aggregation = {"outgoing_mins_per_month" : "mean",
               "outgoing_sms_per_month" : "mean",
               "monthly_mb" : "mean",
               "use_id" : "count"
              }

result.groupby("manufacturer").agg(aggregation)

outgoing_mins_per_month  outgoing_sms_per_month    monthly_mb  \
manufacturer                                                                  
HTC                        299.842955               93.059318   5144.077955   
Huawei                      81.526667                9.500000   1561.226667   
LGE                        111.530000               12.760000   1557.330000   
Lava                        60.650000              261.900000  12458.670000   
Lenovo                     215.920000               12.930000   1557.330000   
Motorola                    95.127500               65.666250   3946.500000   
OnePlus                    354.855000               48.330000   6575.410000   
Samsung                    191.010093               92.390463   4017.318889   
Sony                       177.315625               40.176250   3212.000625   
Vodafone                    42.750000               46.830000   5191.120000   
ZTE                         42.750000               46.830000   5191.120000   

              use_id  
manufacturer          
HTC               44  
Huawei             3  
LGE                2  
Lava               2  
Lenovo             2  
Motorola          16  
OnePlus            6  
Samsung          108  
Sony              16  
Vodafone           1  
ZTE                1